specs/protocol/safer-safes.md

```diff
@@ -0,0 +1,559 @@
+
+#### `getAllScheduledTransactions`
+
+Called by anyone, returns the list of all scheduled but not cancelled or executed transactions for a given safe.
+
+- MUST NOT revert
+- MUST return the list of all scheduled but not cancelled or executed transactions for the given `safe`.
+
+#### `configureTimelockGuard`
+
+Configure the contract as a timelock guard by setting the `timelock_delay`.
+
+- MUST allow an arbitrary number of `safe` contracts to use the contract as a guard.
+- MUST revert if the contract is not enabled as a guard for the `safe`.
+- MUST revert if `timelock_delay` is longer than 1 year.
+- MUST set the caller as a `safe`.
+- MUST take `timelock_delay` as a parameter and store is as related to the `safe`.
+- MUST set the `cancellation_threshold` to 1.
+- MUST emit a `GuardConfigured` event with at least `timelock_delay` as a parameter.
+
+#### `clearTimelockGuard`
+
+Remove the timelock guard configuration by a previously enabled `safe`.
+
+- MUST revert if the contract is enabled as a guard for the `safe`.
+- MUST revert if the contract is not configured for the `safe`.
+- MUST erase the existing `timelock_delay` data related to the calling `safe`.
+- MUST emit a `GuardCleared` event.
+
+#### `scheduleTransaction`
+
+Called by anyone using signatures from Safe owners, registers a transaction in the TimelockGuard for execution after
+the `timelock_delay`.
+
+- MUST revert if the contract is not enabled as a guard for the `safe`.
+- MUST revert if the contract is not configured for the `safe`.
+- MUST take the same parameters as `execTransaction`.
+- MUST revert if an identical transaction has already been scheduled.
+- MUST revert if an identical transaction was cancelled.
+- MUST set `execution_time(safe, txHash)` to `block.timestamp + timelock_delay(safe)`.
+- MUST emit a `TransactionScheduled` event, with at least `safe` and relevant transaction data.
+
+#### `cancelTransaction`
+
+Makes a scheduled transaction not executable. To do so, it builds a no-op transaction at the same nonce as the
+scheduled transaction, and verifies that the supplied signatures for such a no-op transaction are valid and amount to
+`cancellation_threshold(safe)`. If successful, it increases the `cancellation_threshold(safe)` by 1.
+
+- MUST revert if the contract is not enabled as a guard for the `safe`.
+- MUST revert if the contract is not configured for the `safe`.
+- MUST revert if `sum(rejecting_owners(safe, tx)) < cancellation_threshold(safe)`.
+- MUST emit a `TransactionCancelled` event, with at least `safe` and a transaction identifier.
+
+#### `checkTransaction`
+
+Can be called by anyone. It will be called by the `safe` in `execTransaction` if the contract is enabled as a guard.
+It verifies if a given transaction was scheduled and the delay period has passed, and reverts if not.
+
+- MUST revert if the contract is not enabled as a guard for the `safe`.
+- MUST revert if the contract is not configured for the `safe`.
```

Comment:
While implementing this requirement, I realized that it makes it impossible to configure the guard. 
ie. if the installation path is meant to be:
1. enable guard
2. configure guard

Then 2 is not possible with this requirement.

```suggestion
- MUST immediately execute the transaction if the contract is not configured for the `safe`.
```

Reply:
Removed the requirement and added a note at the end. Maybe we need some better explanations around the fact the enabling/disabling the guard in the safe doesn't affect operations, and that setting the timelock to zero or a value different than zero is what really enables/disables it.

Also, `clearTimelockGuard` is the same as `configureTimelockGuard` with a delay of zero.

And I'm not sure if we want to keep the `.configured` field in the struct, or radically simplify and remove both that struct field and the `clearTimelockGuard` function.